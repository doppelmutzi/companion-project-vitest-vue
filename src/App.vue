<template>
  <h1>My awesome dashboard</h1>
  <img :src="imageUrl" :alt="imgAlt" />
  <Counter />
  <TodoFromStore />
  <TodoFromComposable />
</template>

<script setup lang="ts">
import { onMounted, ref } from "vue";
import { useDashboardStore } from "./store";
import TodoFromStore from "./components/TodoFromStore.vue";
import TodoFromComposable from "./components/TodoFromComposable.vue";
import Counter from "../vitest-cheat-sheets/lifecycle-compostion-api/Counter.vue";

const dashboardStore = useDashboardStore();

const imageUrl = ref("");
const imgAlt = ref("");

onMounted(async () => {
  const blob = await dashboardStore.createQuoteImageWithComposable();
  if (blob) {
    imageUrl.value = URL.createObjectURL(blob);
    imgAlt.value = dashboardStore.shortenedQuote;
  }
});
</script>
<style lang="scss"></style>
