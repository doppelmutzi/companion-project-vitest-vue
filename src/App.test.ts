import { flushPromises, shallowMount } from "@vue/test-utils";
import { describe, expect, it, vi } from "vitest";
import App from "./App.vue";

import { createPinia, defineStore } from "pinia";

describe("App", () => {
  it("renders the image correctly", async () => {
    // mock URL.createObjectURL
    vi.stubGlobal("URL", { createObjectURL: () => "a nice URL" });
    // Create a mock store
    const useMockDashboardStore = defineStore("dashboard", () => ({
      createQuoteImageWithComposable: async () => {
        // Create a dummy blob
        const dummyBlob = new Blob();
        return Promise.resolve(dummyBlob);
      },
      shortenedQuote: "Dummy shortened quote",
    }));
    const pinia = createPinia();
    useMockDashboardStore(pinia);

    const wrapper = shallowMount(App, {
      global: {
        plugins: [pinia],
      },
    });

    // make sure to invoke onMounted lifecycle hook and resolve the promise
    await flushPromises();

    const imgEl = wrapper.find("img");
    expect(imgEl.attributes().alt).toBe("Dummy shortened quote");
    expect(imgEl.attributes().src).toBe("a nice URL");

    expect(wrapper.html()).toMatchSnapshot();
  });
});
