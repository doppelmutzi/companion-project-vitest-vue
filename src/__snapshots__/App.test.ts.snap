// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders the image correctly 1`] = `
"<h1>My awesome dashboard</h1>
<img src="a nice URL" alt="Dummy shortened quote">
<todo-from-store-stub></todo-from-store-stub>
<todo-from-composable-stub></todo-from-composable-stub>"
`;
